src/utils.py

```diff
@@ -1872,6 +1872,141 @@ def insert_text(
+        archive: Archive object pointing to locations of used fonts or images
+        rotate: (int) rotate the text with the box by a multiple of 90 degrees.
+        oc: (int) the xref of an OCG / OCMD (Optional Content).
+        overlay: (bool) put text on top of page content.
+        morph: (sequence) (point, matrix) transform result with a matrix
+               using a fixed point.
+    Returns:
+        A float indicating the unused part of the rectangle. If negative, the
+        rectangle was too small for the text.
+    """
+
+    def story_maker(clip, text, adjust=True, rotate=0, css=css, archive=archive):
+        links = []
+
+        def collect(elpos):
+            if not elpos.open_close & 2 or elpos.href is None:
+                return
+            links.append((elpos.href, elpos.rect))
+            return
+
+        rect = fitz.Rect(clip)  # copy to rect
+        # shift to top-left = (0, 0)
+        rect += (-clip.x0, -clip.y0, -clip.x0, -clip.y0)
+        if rotate in (90, 270):  # swap width height
+            rect.x1, rect.y1 = rect.y1, rect.x1
+        orig_height = rect.y1  # original relevant height
+        mycss = "body {margin:1px;}" + css  # append user CSS
+        if isinstance(text, str):  # if a string, convert to a Story
+            story = fitz.Story(html=text, user_css=mycss, archive=archive)
+        else:
+            story = text  # assume we have a story
+
+        # ---------------------------------------------------------------------
+        # Loop for fitting text into rectangle.
+        # We will expand the rectangle until we have a fit in any case. Only
+        # finally we will check "adjust" and report the overflow amount if
+        # no adjustment was intended.
+        # ---------------------------------------------------------------------
+        more = True
+        while more:
+            story.reset()  # rewind story to the beginning
+            fp = io.BytesIO()  # use for file output
+            writer = fitz.DocumentWriter(fp)
+            dev = writer.begin_page(rect)  # make a temporary page
+            more, filled = story.place(rect)  # position text in the rect
+            if more:  # text did not fit in this clip, so enlarge and try again
+                rect *= 1.01  # enlarge by 1%
+                links = []
+                continue
+            # we stayed inside rect
+            # invoke position collector for identifying links
+            story.element_positions(collect, args={})
+            story.draw(dev)
+            writer.end_page()  # end the page
+            writer.close()  # close the writer
+            break  # leave the loop
+
+        try:
+            doc = fitz.open("pdf", fp)  # try make a PDF from memory
+        except Exception as e:
+            # if Document creation failed, return unsuccessful insert
+            print(e)
+            return None, fitz.FZ_MAX_INF_RECT, fitz.Identity, []
```

Comment:
Could we remove the `try..catch`` here and simply let any exception be handled by our caller?